src/Morphader.cpp

```diff
@@ -0,0 +1,324 @@
+}
+
+// equal power crossfade, -1 <= p <= 1
+template <typename T>
+inline T equalPowerCrossfade(T a, T b, const float p) {
+	//return std::min(std::exp(4.f * p), 1.f) * b + std::min(std::exp(4.f * -p), 1.f) * a;
+	return std::min(exponentialBipolar80Pade_5_4(p + 1), 1.f) * b + std::min(exponentialBipolar80Pade_5_4(1 - p), 1.f) * a;
+}
+
+// TExponentialSlewLimiter doesn't appear to work as is required for this application.
+// I think it is due to the absence of the logic that stops the output rising / falling too quickly,
+// i.e. faster than the original signal? I think the following modification would yield the
+// expected behaviour (see 2 lines in process).
+/*
+template <typename T = float>
+struct TExponentialSlewLimiter {
+	T out = 0.f;
+	T riseLambda = 0.f;
+	T fallLambda = 0.f;
+
+	void reset() {
+		out = 0.f;
+	}
+
+	void setRiseFall(T riseLambda, T fallLambda) {
+		this->riseLambda = riseLambda;
+		this->fallLambda = fallLambda;
+	}
+	T process(T deltaTime, T in) {
+		// MODIFICATION:
+		T rising = in > out;
+		T lambda = simd::ifelse(rising, riseLambda, fallLambda);
+		T y = out + (in - out) * lambda * deltaTime;
+		// If the change from the old out to the new out is too small for floats, set `in` directly.
+		out = simd::ifelse(out == y, in, y);
+
+		// MODIFICATION:
+		out = simd::ifelse(rising, simd::ifelse(out > in, in, y), simd::ifelse(out < in, in, y));
+		return out;
+	}
+	DEPRECATED T process(T in) {
+		return process(1.f, in);
+	}
+};
+*/
+
+// For now, I provide this implementation (essentialy the same as SlewLimiter.cpp), but ideally I
+// would replace with updated library function
+struct ExpLogSlewLimiter {
+
+	float out = 0.f;
+	float slew = 0.f;
+
+	void reset() {
+		out = 0.f;
+	}
+
+	void setSlew(float slew) {
+		this->slew = slew;
+	}
+	float process(float deltaTime, float in) {
+		if (in > out) {
+			out += slew * (in - out) * deltaTime;
+			if (out > in) {
+				out = in;
+			}
+		}
+		else if (in < out) {
+			out += slew * (in - out) * deltaTime;
+			if (out < in) {
+				out = in;
+			}
+		}
+		return out;
+	}
+};
+
+
+struct Morphader : Module {
+	enum ParamIds {
+		CV_PARAM,
+		ENUMS(A_LEVEL, 4),
+		ENUMS(B_LEVEL, 4),
+		ENUMS(MODE, 4),
+		FADER_LAG_PARAM,
+		FADER_PARAM,
+		NUM_PARAMS
+	};
+	enum InputIds {
+		ENUMS(CV_INPUT, 4),
+		ENUMS(A_INPUT, 4),
+		ENUMS(B_INPUT, 4),
+		NUM_INPUTS
+	};
+	enum OutputIds {
+		ENUMS(OUT, 4),
+		NUM_OUTPUTS
+	};
+	enum LightIds {
+		ENUMS(A_LED, 4),
+		ENUMS(B_LED, 4),
+		NUM_LIGHTS
+	};
+	enum CrossfadeMode {
+		AUDIO_MODE,
+		CV_MODE
+	};
+
+	static const int NUM_MIXER_CHANNELS = 4;
+	const simd::float_4 normal10VSimd = {10.f};
+	ExpLogSlewLimiter slewLimiter;
+
+	// minimum and maximum slopes in volts per second, they specify the time to get
+	// from A (-1) to B (+1)
+	constexpr static float slewMin = 2.0 / 15.f;
+	constexpr static float slewMax = 2.0 / 0.01f;
+
+	struct AudioCVModeParam : ParamQuantity {
+		std::string getDisplayValueString() override {
```

Comment:
This should be replaced with `configSwitch()` in Rack 2.

Reply:
Yeah there's a load of boilerplate I'm looking forward to removing :)